mergify_engine/__init__.py

```diff
@@ -48,3 +55,64 @@ def fixup_sentry_reporting(event, hint):  # pragma: no cover
         before_send=fixup_sentry_reporting,
         integrations=[CeleryIntegration(), FlaskIntegration(), RedisIntegration()],
     )
+
+
+RETRY = urllib3.Retry(
+    total=None,
+    redirect=3,
+    connect=5,
+    read=5,
+    status=5,
+    backoff_factor=0.2,
+    status_forcelist=list(range(500, 599)) + [429],
+    method_whitelist=[
+        "HEAD",
+        "TRACE",
+        "GET",
+        "PUT",
+        "OPTIONS",
+        "DELETE",
+        "POST",
+        "PATCH",
+    ],
+    raise_on_status=False,
+)
+
+
+SINGLE_PULL_API_RE = re.compile(
+    f"^https://api.{config.GITHUB_DOMAIN}:443/repos/[^/]+/[^/]+/pulls/\\d+$"
+)
+
+real_session_init = requests.sessions.Session.__init__
+
+
+class CustomCacheControlAdapter(cachecontrol.CacheControlAdapter):
+    def send(self, request, *args, **kwargs):
+        # NOTE(sileht): never ever cache get_pull(), we need mergeable_state
+        # up2date and when it changes etag is not updated...
+        if SINGLE_PULL_API_RE.match(request.url):
+            cache_url = self.controller.cache_url(request.url)
+            self.cache.delete(cache_url)
+        return super().send(request, *args, **kwargs)
+
+
+def cached_session_init(self, *args, **kwargs):
+    real_session_init(self, *args, **kwargs)
+
+    dashboard_adapter = requests.adapters.HTTPAdapter(max_retries=RETRY)
+    self.mount(config.SUBSCRIPTION_URL, dashboard_adapter)
+    if config.WEBHOOK_APP_FORWARD_URL:
+        self.mount(config.WEBHOOK_APP_FORWARD_URL, dashboard_adapter)
+    if config.WEBHOOK_MARKETPLACE_FORWARD_URL:
+        self.mount(config.WEBHOOK_MARKETPLACE_FORWARD_URL, dashboard_adapter)
```

Comment:
hm why do we need a not-cached adapter for this one?

Reply:
Cached adapter works only if API server support `etag`, our dashboard api doesn'. So, it's useless.

Reply:
Is it just useless, as is it has no impact, or it does not work?
I'm just wondering if it's worth having this 4 lines.

Reply:
I double checked cachecontrol, that should be safe, only 301 and 304 code have special handling, but client that doesn't send etag should never received them, so.

Reply:
> but client that doesn't send etag should never received them, so.

Ok, that makes sense then!